[optimize](compaction) optimize cu compaction policy in pick rowset

yixiutt opened the pull request:
origin _levels is between size_based_compaction_lower_bound_size and
size_based_promotion_size, default conf is 64M and 1024M, which is too
big for cu compaction, and once rowset size bigger than 64M it will be
promoted to base, so this rule seems to be useless.

update _leves to between 32M and 1K, which will reduce a lot of cpu.
In our test, we have 20 concurrent stream load and each rowset in disk
is 3K, after introduce this change, we save about 25% cpu.

# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

Commit:
[optimize](compaction) optimize cu compaction policy in pick rowset

origin _levels is between size_based_compaction_lower_bound_size and
size_based_promotion_size, default conf is 64M and 1024G, which is too
big for cu compaction, and once rowset size bigger than 64M it will be
promoted to base, so this rule seems to be useless.

update _leves to 32M and 1K, which will reduce a lot of cpu.
In our test, we have 20 concurrent stream load and each rowset in disk
is 3K, after introduce this change, we save about 25% cpu.

Commit:
[optimize](compaction) optimize cu compaction policy in pick rowset

origin _levels is between size_based_compaction_lower_bound_size and
size_based_promotion_size, default conf is 64M and 1024G, which is too
big for cu compaction, and once rowset size bigger than 64M it will be
promoted to base, so this rule seems to be useless.

update _leves to 32M and 1K, which will reduce a lot of cpu.
In our test, we have 20 concurrent stream load and each rowset in disk
is 3K, after introduce this change, we save about 25% cpu.

zhannngchen commented on be/src/olap/cumulative_compaction_policy.cpp:
```diff
@@ -32,10 +32,11 @@ SizeBasedCumulativeCompactionPolicy::SizeBasedCumulativeCompactionPolicy(
+    int64_t i_size = size_based_compaction_lower_bound_size / 2;
 
-    while (i_size >= size_based_compaction_lower_bound_size) {
+    while (i_size >= 1024) {
```
This value should be configurable?

BiteTheDDDDt committed:
[bugfix] signed long will be converted to signed long during dcheck and cause dcheck fail (#10047)

hf200012 committed:
[fix]Document image path error (#10050)

Document image path error

morningman committed:
[improvement][refactor](image) refactor the read and load method of meta image #10005

BiteTheDDDDt committed:
[Bug] [Vectorized] code dump on aggregate node over union node (#10040)

* miss check passthrough on vectorized

* format and add test

* update

yinzhijian committed:
[fix](fe) select stmt will make BE coredump when its castExpr is like cast(int as array<>) (#9995)

* [fix](fe) select stmt will make BE coredump when its castExpr is like cast(int as array<>)

* fix implicit cast scalar type bug

* Revert "fix implicit cast scalar type bug"

This reverts commit 1f05b6bab72430214dca88f386b50ef9a081e60a.

* only check array cast, retrigger

yangzhg committed:
[dependency][enhancement] support build libhdfs in arm cpus (#10018)

Supports native hdfs functionality on arm cpu
This pr mainly upgrades libdfs3 and supports running on arm，and make libhdfs3 with kerberos as default

qzsee committed:
[fix](planner) fix don't rewrite nested union statement bug (#8513)

Issue Number: close #8512

carlvinhust2012 committed:
[metric] add some metrics for cpu and memory (#9887)

1. add some metrics for cpu monitor;
2. add metrics for process state monitor;
3. add metrics for memory monitor;

It is convenient for us to use grafana to filter through different conditions.

After the added, we can find the cpu metrics like this：
doris_be_cpu{device="cpu1",mode="guest_nice"} 0
doris_be_cpu{device="cpu1",mode="guest"} 0
doris_be_cpu{device="cpu1",mode="steal"} 0
doris_be_cpu{device="cpu1",mode="soft_irq"} 107168
doris_be_cpu{device="cpu1",mode="irq"} 0
doris_be_cpu{device="cpu1",mode="iowait"} 3726931
doris_be_cpu{device="cpu1",mode="idle"} 2358039214
doris_be_cpu{device="cpu1",mode="system"} 58699464
doris_be_cpu{device="cpu1",mode="nice"} 1700438
doris_be_cpu{device="cpu1",mode="user"} 54974091

we can find the memory metrics as follow：
doris_be_memory_pswpin 167785
doris_be_memory_pswpout 203724
doris_be_memory_pgpgin 22308762092
doris_be_memory_pgpgout 152101956232


we also can find the process metrics as follow:
doris_be_proc{mode="interrupt"} 421721020416
doris_be_proc{mode="ctxt_switch"} 2806640907317
doris_be_proc{mode="procs_running"} 8
doris_be_proc{mode="procs_blocked"} 3

morrySnow committed:
[fix](planner) produce wrong result when use bucket shuffle join with colocate left table (#10045)

When plan bucket shuffle join, we need to know left table bucket number.
Currently, we use tablet number directly based on the assumption that left table has only one partition.
But, when left table is colocated table, it could have more than one partition.
In this case, some data in right table will be dropped incorrectly and produce wrong result for query.

reproduce could follow regression test in PR.

liutang123 committed:
[fix] Do not send drop task when replay drop table (#10062)

When doing checkpoint, FE will sends DropTask to BE.
This PR prohibit this conduct.

zy-kkk committed:
[test] Add drop force to cases associated with ALTER operations (#10049)

adonis0147 committed:
[feature-wip](array-type) Support array type which doesn't contain null (#9809)

mrhhsg committed:
[Enhancement]Decode bitshuffle data before adding it into PageCache (#10036)

* [Enhancement]Decode bitshuffle data before add into PageCache

* Fix be ut failed

Gabriel39 committed:
[BUGFIX] Fix wrong column types in result file sink (#10079)

BePPPower committed:
[doc] add description ofd[DEFAULT CURRENT_TIMESTAMP] (#10053)

Jibing-Li committed:
[feature](multi-catalog) Change DatabaseIf APIs' return type to TableIf. (#10044)

Change the DatabaseIf APIs' return type to TableIf.
Use generics in DatabaseIf, to avoid changing the return type in Database.
Currently Database class use type Table, I'm try to avoid changing it to TableIf.
Because in this case, we need to change a lot of code.

Gabriel39 committed:
[Improvement] add regression test for outfile (#10058)

xinyiZzz committed:
[fix](brpc) Embed serialized request into the attachment and transmit it through http brpc (#9803)

When the length of `Tuple/Block data` is greater than 2G, serialize the protoBuf request and embed the
`Tuple/Block data` into the controller attachment and transmit it through http brpc.

This is to avoid errors when the length of the protoBuf request exceeds 2G:
`Bad request, error_text=[E1003]Fail to compress request`.

In #7164, `Tuple/Block data` was put into attachment and sent via default `baidu_std brpc`,
but when the attachment exceeds 2G, it will be truncated. There is no 2G limit for sending via `http brpc`.

Also, in #7921, consider putting `Tuple/Block data` into attachment transport by default, as this theoretically
reduces one serialization and improves performance. However, the test found that the performance did not improve,
but the memory peak increased due to the addition of a memory copy.

jacktengg committed:
[improvement] send merged runtime filter asynchrously (#10080)

zxealous committed:
[typo] Fix typos in comments (#10106)

Gabriel39 committed:
[Improvement] set table name in olap scanner (#10102)

yangzhg committed:
[deps] update libhdfs3 to fix a uuid set problem (#10092)

yangzhg committed:
[enhancement] support simd instructions on arm cpus through sse2neon (#10068)

* [enhancement] support simd instructions on arm cpus through sse2neon

BiteTheDDDDt committed:
[build] use inline to replace static (#10087)

caoliang-web committed:
[docs] Add common error messages to doris backup (#10048)

yinzhijian committed:
[typo] Fix typos in runtime_state.cpp (#10112)

yinzhijian committed:
[typo] Fix typos in comments (#10111)

yinzhijian committed:
[spell] Fix spell error in row_batch.h (#10109)

cambyzju committed:
[doc] Add docs for SHOW TABLETS (#10105)

* add docs for SHOW TABLETS

* update

* add more examples for SHOW TABLETS

Co-authored-by: cambyzju <zhuxiaoli01@baidu.com>

BiteTheDDDDt committed:
[bugfix]fix segmentation fault at unalign address cast to int128 (#10094)

morningman committed:
[fix] fix bug that show proc "/cluster_balance/history_tablets" return malformat error (#10073)

morrySnow committed:
[fix](regression)bucket shuffle join with collocate table should use order_qt (#10082)

HappenLee committed:
[Bug] Fix bug push value predicate of unique table when have sequence column (#10060)

Co-authored-by: lihaopeng <lihaopeng@baidu.com>

qzsee committed:
[Enhancement] (Nereids) scalar expression rewrite framework (#9942)

Issue Number: close #9633

The scalar expression is rewritten using the visitor pattern as a traversal.

In the abstract class ExpressionVisitor, which contains all predicate to rewrite.

We have provided a rewrite rules interface ExpressionRewriteRule, AbstractExpressionRewriteRule class implements the interface and expanded the ExpressionVisitor, if we want to realize an expression rewriting rules, Direct implementation AbstractExpressionRewriteRule provided in the method of traversing the predicate.

There are two rules to refer: NormalizeExpressionRule and SimplifyNotExprRule

HappenLee committed:
[Bug] return type is not always nullable of function (#10116)

Co-authored-by: lihaopeng <lihaopeng@baidu.com>

englefly committed:
[test] tpch q3 rewrite, change join order, make lineitem on left side (#10055)

rewrite the sql in tpch test tools

Kikyou1997 committed:
[feature](nereids) Plan Translator (#9993)

Issue Number: close #9621

Add following physical operator: PhysicalAgg PhysicalSort PhysicalHashJoin

Add basic logic of plan translator

1. add new agg phase enum for nereids
2. remove the Analyzer from PlanContext.java
3. implement PlanTranslator::visitPhysicalFilter

gtchaos committed:
[feature] Support read hive external table and outfile into HDFS that authenticated by kerberos (#9579)

At present, Doris can only access the hadoop cluster with kerberos authentication enabled by broker, but Doris BE itself 
does not supports access to a kerberos-authenticated HDFS file.

This PR hope solve the problem.

When create hive external table, users just specify following properties to access the hdfs data with kerberos authentication enabled:

```sql
CREATE EXTERNAL TABLE t_hive (
k1 int NOT NULL COMMENT "",
k2 char(10) NOT NULL COMMENT "",
k3 datetime NOT NULL COMMENT "",
k5 varchar(20) NOT NULL COMMENT "",
k6 double NOT NULL COMMENT ""
) ENGINE=HIVE
COMMENT "HIVE"
PROPERTIES (
'hive.metastore.uris' = 'thrift://192.168.0.1:9083',
'database' = 'hive_db',
'table' = 'hive_table',
'dfs.nameservices'='hacluster',
'dfs.ha.namenodes.hacluster'='n1,n2',
'dfs.namenode.rpc-address.hacluster.n1'='192.168.0.1:8020',
'dfs.namenode.rpc-address.hacluster.n2'='192.168.0.2:8020',
'dfs.client.failover.proxy.provider.hacluster'='org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider',
'dfs.namenode.kerberos.principal'='hadoop/_HOST@REALM.COM'
'hadoop.security.authentication'='kerberos',
'hadoop.kerberos.principal'='doris_test@REALM.COM',
'hadoop.kerberos.keytab'='/path/to/doris_test.keytab'
);
```

If you want  to `select into outfile` to HDFS that kerberos authentication enable, you can refer to the following SQL statement：

```sql
select * from test into outfile "hdfs://tmp/outfile1" 
format as csv
properties
(
'fs.defaultFS'='hdfs://hacluster/',
'dfs.nameservices'='hacluster',
'dfs.ha.namenodes.hacluster'='n1,n2',
'dfs.namenode.rpc-address.hacluster.n1'='192.168.0.1:8020',
'dfs.namenode.rpc-address.hacluster.n2'='192.168.0.2:8020',
'dfs.client.failover.proxy.provider.hacluster'='org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider',
'dfs.namenode.kerberos.principal'='hadoop/_HOST@REALM.COM'
'hadoop.security.authentication'='kerberos',
'hadoop.kerberos.principal'='doris_test@REALM.COM',
'hadoop.kerberos.keytab'='/path/to/doris_test.keytab'
);
```

xinyiZzz committed:
[fix](mem tracker) Fix some memory leaks, inaccurate statistics, core dump, deadlock bugs (#10072)

1. Fix the memory leak. When the load task is canceled, the `IndexChannel` and `NodeChannel` mem trackers cannot be destructed in time.
2. Fix Load task being frequently canceled by oom and inaccurate `LoadChannel` mem tracker limit, and rewrite the variable name of `mem limit` in `LoadChannel`.
3. Fix core dump, when logout task mem tracker, phmap erase fails, resulting in repeated logout of the same tracker.
4. Fix the deadlock, when add_child_tracker mem limit exceeds, calling log_usage causes `_child_trackers_lock` deadlock.
5. Fix frequent log printing when thread mem tracker limit exceeds, which will affect readability and performance.
6. Optimize some details of mem tracker display.

jackwener committed:
[enhancement](community): polish doc to reformat (#10137)

hf200012 committed:
[doc]Added be enable_stream_load_record configuration description (#10130)

JNSimba committed:
[modify default config]add be 2pc config  enbale  defalut (#10110)

Co-authored-by: wudi <>

wangyf0555 committed:
[doc][website]update home page content and add slack button (#10091)

* fix docs bugs with sidebar can not display and some style problems

platoneko committed:
[fix] Fix the bug that data balance causes tablet loss (#9971)

1. Provide a FE conf to test the reliability in single replica case when tablet scheduling are frequent.
2. According to #6063, almost apply this fix on current code.

zhangstar333 committed:
[Vectorized][UDF] support java-udaf (#9930)

pengxiangyu committed:
Add storage policy for remote storage migration (#9997)

yinzhijian committed:
[test] add cast array regression test (#10069)

* [test] add cast array regression test

924060929 committed:
[Enhancement][Refactor](Nereids) generate pattern by operator and refactor Plan and NODE_TYPE generic type (#10019)

This pr support
1. remove the generic type from operator, remove some NODE_TYPE from plan and expression
2. refactor Plan and NODE_TYPE generic type
3. support child class matching by TypePattern
4. analyze the code of operator and generate pattern makes it easy to create rules.


e.g. 
```java
class LogicalJoin extends LogicalBinaryOperator;
class PhysicalFilter extends PhysicalUnaryOperator;
```

will generate the code
```java
interface GeneratedPatterns extends Patterns {
  default PatternDescriptor<LogicalBinaryPlan<LogicalJoin, Plan, Plan>, Plan> logicalJoin() {
      return new PatternDescriptor<LogicalBinaryPlan<LogicalJoin, Plan, Plan>, Plan>(
          new TypePattern(LogicalJoin.class, Pattern.FIXED, Pattern.FIXED),
          defaultPromise()
      );
  }
  
  default <C1 extends Plan, C2 extends Plan>
  PatternDescriptor<LogicalBinaryPlan<LogicalJoin, C1, C2>, Plan>
          logicalJoin(PatternDescriptor<C1, Plan> child1, PatternDescriptor<C2, Plan> child2) {
      return new PatternDescriptor<LogicalBinaryPlan<LogicalJoin, C1, C2>, Plan>(
          new TypePattern(LogicalJoin.class, child1.pattern, child2.pattern),
          defaultPromise()
      );
  }

  default PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, Plan>, Plan> physicalFilter() {
      return new PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, Plan>, Plan>(
          new TypePattern(PhysicalFilter.class, Pattern.FIXED),
          defaultPromise()
      );
  }
  
  default <C1 extends Plan>
  PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, C1>, Plan>
          physicalFilter(PatternDescriptor<C1, Plan> child1) {
      return new PatternDescriptor<PhysicalUnaryPlan<PhysicalFilter, C1>, Plan>(
          new TypePattern(PhysicalFilter.class, child1.pattern),
          defaultPromise()
      );
  }
}
```
and then we don't have to add pattern for new operators.

this function utilizing jsr269 to do something in compile time, and utilizing antlr4 to analyze the code of `Operator`, then we can generate corresponding pattern.


pattern generate steps:
1. maven-compiler-plugin in the pom.xml will compile fe-core three terms. first term will compile `PatternDescribable.java` and `PatternDescribableProcessor.java`
2. second compile term will compile `PatternDescribableProcessPoint.java`, and enable annotation process `PatternDescribableProcessor`, PatternDescribableProcessor will receive the event and know that `PatternDescribableProcessPoint` class contains the `PatternDescribable` annotation.
3. `PatternDescribableProcessor` will not process `PatternDescribableProcessPoint`, but find all java file exists in `operatorPath` that specify in pom.xml, and then parse to Java AST(abstract syntax tree).
5. PatternDescribableProcessor collect java AST and use `PatternGeneratorAnalyzer` to analyze AST, find the child class file for `PlanOperator` then generate `GeneratedPatterns.java` by the AST.
6. third compile term will compile `GeneratedPatterns.java` and other java file.

wangyf0555 committed:
[doc](website)modify home page text and navbar (#10148)

* fix docs bugs with sidebar can not display and some style problems

Commit:
Merge branch 'cu_compaction_policy_bugfix_yixiu' of https://github.com/yixiutt/incubator-doris into cu_compaction_policy_bugfix_yixiu
